File: lingobot-backend/src/main/java/com/lingobot/core/user/auth/service/AuthService.java
package com.lingobot.core.user.auth.service;

import com.lingobot.core.user.auth.dto.AuthResponse;
import com.lingobot.core.user.auth.dto.ChangePasswordRequest;
import com.lingobot.core.user.auth.dto.LoginWithCodeRequest;
import com.lingobot.core.user.auth.dto.RegisterWithCodeRequest;
import com.lingobot.core.user.auth.dto.SendLoginCodeRequest;
import com.lingobot.core.user.auth.dto.UserDTO;

/**
 * 认证服务接口。
 *
 * 定义用户认证相关的核心操作：注册、登录、验证码发送、账户管理等。
 * 支持邮箱验证码注册、邮箱密码验证码登录，包含登录失败次数限制、IP 封锁等安全机制。
 */
public interface AuthService {
    // 用户注册（邮箱 + 验证码方式）
    AuthResponse registerWithCode(RegisterWithCodeRequest request, String clientIp);
    // 发送登录验证码（需先验证邮箱密码）
    void sendLoginVerificationCode(SendLoginCodeRequest request, String clientIp);
    // 用户登录（邮箱 + 密码 + 验证码方式）
    AuthResponse loginWithCode(LoginWithCodeRequest request, String clientIp);
    // 获取当前登录用户的完整信息
    UserDTO getCurrentUser();
    // 获取当前登录用户的 ID（未登录时返回 null）
    Long getCurrentUserId();
    // 用户退出登录，清除安全上下文
    void logout();
    // 注销当前用户账户，删除用户数据
    void deactivateAccount();
    // 修改当前用户的密码
    void changePassword(ChangePasswordRequest request);
    // 更新当前用户的头像
    void updateAvatar(String avatarBase64);
    // 更新当前用户的昵称，返回新的认证响应（包含新 Token）
    AuthResponse updateUsername(String username);
}

File: lingobot-frontend/src/services/authService.ts
import { httpClient, authUtils } from './httpClient';
import { 
  UserDTO, 
  AuthResponse, 
  LoginRequest, 
  RegisterRequest, 
  RegisterWithCodeRequest,
  SendLoginCodeRequest,
  LoginWithCodeRequest
} from '../types';

export const authService = {
  login: async (request: LoginRequest): Promise<AuthResponse> => {
    const response = await httpClient.postRaw('/auth/login', request);
    const result = await response.json();
    const data: AuthResponse = result.data;
    
    const user: UserDTO = {
      id: data.userId,
      username: data.username,
      role: data.role,
      avatar: data.avatar,
      createdAt: new Date().toISOString(),
      balance: data.balance ? Number(data.balance) : 0
    };
    
    authUtils.setAuth(data.token, user);
    window.dispatchEvent(new CustomEvent('auth:login', { detail: { user } }));
    
    return data;
  },
  
  sendLoginCode: async (request: SendLoginCodeRequest): Promise<void> => {
    await httpClient.post('/auth/send-login-code', request);
  },
  
  loginWithCode: async (request: LoginWithCodeRequest): Promise<AuthResponse> => {
    const response = await httpClient.postRaw('/auth/login-with-code', request);
    const result = await response.json();
    const data: AuthResponse = result.data;
    
    const user: UserDTO = {
      id: data.userId,
      username: data.username,
      email: data.email,
      role: data.role,
      avatar: data.avatar,
      createdAt: new Date().toISOString(),
      balance: data.balance ? Number(data.balance) : 0
    };
    
    authUtils.setAuth(data.token, user);
    window.dispatchEvent(new CustomEvent('auth:login', { detail: { user } }));
    
    return data;
  },
  
  register: async (request: RegisterRequest): Promise<AuthResponse> => {
    const response = await httpClient.postRaw('/auth/register', request);
    const result = await response.json();
    const data: AuthResponse = result.data;
    
    const user: UserDTO = {
      id: data.userId,
      username: data.username,
      email: data.email,
      role: data.role,
      avatar: data.avatar,
      createdAt: new Date().toISOString(),
      balance: data.balance ? Number(data.balance) : 0
    };
    
    authUtils.setAuth(data.token, user);
    window.dispatchEvent(new CustomEvent('auth:login', { detail: { user } }));
    
    return data;
  },
  
  sendVerificationCode: async (email: string): Promise<void> => {
    await httpClient.post('/auth/send-verification-code', { email });
  },
  
  registerWithCode: async (request: RegisterWithCodeRequest): Promise<AuthResponse> => {
    const response = await httpClient.postRaw('/auth/register-with-code', request);
    const result = await response.json();
    const data: AuthResponse = result.data;
    
    const user: UserDTO = {
      id: data.userId,
      username: data.username,
      email: request.email,
      role: data.role,
      avatar: data.avatar,
      createdAt: new Date().toISOString(),
      balance: data.balance ? Number(data.balance) : 0
    };
    
    authUtils.setAuth(data.token, user);
    window.dispatchEvent(new CustomEvent('auth:login', { detail: { user } }));
    
    return data;
  },
  
  getCurrentUser: async (): Promise<UserDTO> => {
    const user = await httpClient.get<UserDTO>('/auth/me');
    return user;
  },
  
  logout: () => {
    authUtils.clearAuth();
    window.dispatchEvent(new CustomEvent('auth:logout'));
  },
  
  changePassword: async (currentPassword: string, newPassword: string, confirmPassword: string): Promise<void> => {
    await httpClient.post('/auth/change-password', { 
      currentPassword, 
      newPassword, 
      confirmPassword 
    });
  },
  
  deactivateAccount: async (): Promise<void> => {
    await httpClient.post('/auth/deactivate');
    authUtils.clearAuth();
    window.dispatchEvent(new CustomEvent('auth:logout'));
  },
  
  updateAvatar: async (avatar: string): Promise<void> => {
    await httpClient.post('/auth/update-avatar', { avatar });
  },
  
  updateUsername: async (username: string): Promise<AuthResponse> => {
    const response = await httpClient.postRaw('/auth/update-username', { username });
    const result = await response.json();
    const data: AuthResponse = result.data;
    
    const updatedUser: UserDTO = {
      id: data.userId,
      username: data.username,
      email: data.email,
      role: data.role,
      avatar: data.avatar,
      createdAt: new Date().toISOString(),
      balance: data.balance ? Number(data.balance) : 0
    };
    
    authUtils.setAuth(data.token, updatedUser);
    window.dispatchEvent(new CustomEvent('auth:username-updated', { detail: { user: updatedUser } }));
    
    return data;
  }
};
